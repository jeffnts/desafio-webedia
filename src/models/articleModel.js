import mongoose from 'mongoose'
const Schema = mongoose.Schema

const articleSchema = new Schema({
    
})

export default  mongoose.model('Article', userSchema)