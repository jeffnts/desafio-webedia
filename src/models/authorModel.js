import mongoose from 'mongoose'
import mongoosePaginate from 'mongoose-paginate-v2'

const Schema = mongoose.Schema

const authorSchema = new Schema({
    name:{
        type: String,
        required: true
    }
    
})

authorSchema.plugin(mongoosePaginate)

export default  mongoose.model('Author', authorSchema)