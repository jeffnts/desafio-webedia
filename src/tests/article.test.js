//import app from '../app'
const app = require('../app')
const request = require('supertest')(app)
const connection = require('../config/database/testDB')
const {sanitizeTestObject} = require('../config/tests/sanitizeTestObject')

import authorModel from '../models/authorModel'
import articleModel from '../models/articleModel'

const dropArticleCollection = async () => articleModel.remove()
const dropAuthorCollection = async () => authorModel.remove()

beforeAll(async() => {
  dropArticleCollection()
  dropAuthorCollection()
  await request.post('/api/authors').send(authorObject)
})

afterAll(() => {
  dropArticleCollection()
  dropAuthorCollection()
  connection.disconect()
})

let authorObject = {
  "_id": "TESTOBJECT01",
  "name": "Monteiro Lobato"
}

let article = {
  "_id": "TESTOBJECT02",
  "title": "Artigo 1",
  "permalink": "artigo-1",
  "authors":["TESTOBJECT01"]
}
let article2 = {
  "_id": "TESTOBJECT03",
  "title": "Artigo 1",
  "permalink": "artigo-1",
  "authors":["TESTOBJECT11"]
}
let article3 = {
  "_id": "TESTOBJECT03",
  "title": "Artigo 1",
  "permalink": "artigo-1"
}

describe('Article tests', () =>{
  describe('POST method', () =>{
    test('It should create a new article.', async () =>{
      const res = await request.post('/api/articles').send(article)
      expect(sanitizeTestObject(res.body)).toMatchSnapshot()
      expect(res.statusCode).toBe(201)
    })

    test('It should not create a article with a inexistent author.', async () =>{
      const res = await request.post('/api/articles').send(article2)
      expect(sanitizeTestObject(res.body)).toMatchSnapshot()
      expect(res.statusCode).toBe(404)
    })

    test('It should not create a article without a  author.', async () =>{
      const res = await request.post('/api/articles').send(article3)
      expect(sanitizeTestObject(res.body)).toMatchSnapshot()
      expect(res.statusCode).toBe(406)
    })

  })

  describe('GET method', () =>{
    test('It shoul return all articles', async () =>{
      const res = await request.get('/api/articles')
      expect(sanitizeTestObject(res.body)).toMatchSnapshot()
      expect(res.statusCode).toBe(200)
    })

    test('It should return the articles according the permalink passed as parameter', async () =>{
      const res = await request.get(`/api/articles/${article.permalink}`)
      expect(sanitizeTestObject(res.body)).toMatchSnapshot()
      expect(res.statusCode).toBe(200)
    })
    test('It should not return the articles with the wrong permalink passed as parameter', async () =>{
      const res = await request.get(`/api/articles/meu-artigo`)
      expect(sanitizeTestObject(res.body)).toMatchSnapshot()
      expect(res.statusCode).toBe(404)
    })
  })

  describe('PUT method', ()=>{
    test('It should not update the article with the wrong permalink passed as parameter', async () =>{
      const res = await request.put(`/api/articles/meu-artigo`).send({title: 'Artigo de Zé do Livro'})
      expect(sanitizeTestObject(res.body)).toMatchSnapshot()
      expect(res.statusCode).toBe(404)
    })
    test('It should update the article according the permalink passed as parameter', async () =>{
      const res = await request.put(`/api/articles/${article.permalink}`).send({title: 'Artigo de Zé do Livro'})
      expect(sanitizeTestObject(res.body)).toMatchSnapshot()
      expect(res.statusCode).toBe(200)
    })
  })

  describe('DELETE method', ()=>{
    test('It should not delete the article with the wrong permalink passed as parameter', async () =>{
      const res = await request.delete('/api/articles/meu-artigo')
      expect(sanitizeTestObject(res.body)).toMatchSnapshot()
      expect(res.statusCode).toBe(404)
    })
    test('It should delete the article according the permalink passed as parameter', async () =>{
      const res = await request.delete(`/api/articles/${article.permalink}`)
      expect(sanitizeTestObject(res.body)).toMatchSnapshot()
      expect(res.statusCode).toBe(200)
    })

  })
})