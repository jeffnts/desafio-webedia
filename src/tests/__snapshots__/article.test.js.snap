// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article tests DELETE method It should delete the article according the permalink passed as parameter 1`] = `
Object {
  "message": "Artigo deletado com sucesso!",
}
`;

exports[`Article tests DELETE method It should not delete the article with the wrong permalink passed as parameter 1`] = `
Object {
  "message": "Este artigo não existe.",
}
`;

exports[`Article tests GET method It shoul return all articles 1`] = `
Object {
  "articles": Object {
    "docs": Array [
      Object {
        "__v": 0,
        "_id": "ObjectId",
        "authors": Array [
          Object {
            "__v": 0,
            "_id": "ObjectId",
            "name": "Monteiro Lobato",
          },
        ],
        "comments": Array [],
        "permalink": "artigo-1",
        "publicationDate": "2019-01-22T11:18:51.337Z",
        "title": "Artigo 1",
        "updateDate": "2019-01-22T11:18:51.337Z",
      },
    ],
    "limit": "EMPTY",
    "offset": "EMPTY",
    "totalDocs": 1,
  },
}
`;

exports[`Article tests GET method It should return the articles according the id passed as parameter 1`] = `
Object {
  "article": Object {
    "__v": 0,
    "_id": "ObjectId",
    "authors": Array [
      Object {
        "__v": 0,
        "_id": "ObjectId",
        "name": "Monteiro Lobato",
      },
    ],
    "comments": Array [],
    "permalink": "artigo-1",
    "publicationDate": "2019-01-22T11:18:51.337Z",
    "title": "Artigo 1",
    "updateDate": "2019-01-22T11:18:51.337Z",
  },
}
`;

exports[`Article tests GET method It should return the articles according the permalink passed as parameter 1`] = `
Object {
  "article": Object {
    "__v": 0,
    "_id": "ObjectId",
    "authors": Array [
      Object {
        "__v": 0,
        "_id": "ObjectId",
        "name": "Monteiro Lobato",
      },
    ],
    "comments": Array [],
    "permalink": "artigo-1",
    "publicationDate": "2019-01-22T11:26:29.923Z",
    "title": "Artigo 1",
    "updateDate": "2019-01-22T11:26:29.923Z",
  },
}
`;

exports[`Article tests POST method It should create a new article. 1`] = `
Object {
  "message": "Artigo criado com sucesso!",
}
`;

exports[`Article tests POST method It should not create a article with a inexistent author. 1`] = `
Object {
  "message": "Este autor não existe.",
}
`;

exports[`Article tests POST method It should not create a article without a  author. 1`] = `
Object {
  "message": "É obrigatório um artigo ter autor!",
}
`;

exports[`Article tests PUT method It should not update the article with the wrong permalink passed as parameter 1`] = `
Object {
  "message": "Este artigo não existe.",
}
`;

exports[`Article tests PUT method It should update the article according the permalink passed as parameter 1`] = `
Object {
  "message": "Artigo atualizado com sucesso!",
}
`;
